virttest/libvirt_vm.py

```diff
@@ -2954,23 +2954,40 @@ def remove_package(self, name, ignore_status=False):
             logging.error("Removal of package %s failed", name)
         session.close()
 
-    def prepare_guest_agent(self, prepare_xml=True, channel=True, start=True):
+    def prepare_guest_agent(self, prepare_xml=True, channel=True, start=True,
+                            source_path=None, target_name='org.qemu.guest_agent.0'):
         """
         Prepare qemu guest agent on the VM.
 
         :param prepare_xml: Whether change VM's XML
         :param channel: Whether add agent channel in VM. Only valid if
                         prepare_xml is True
         :param start: Whether install and start the qemu-ga service
+        :param source_path: Source path of the guest agent channel
+        :param target_name: Target name of the guest agent channel
         """
         if prepare_xml:
+            if self.is_alive():
+                self.destroy()
             vmxml = libvirt_xml.VMXML.new_from_inactive_dumpxml(self.name)
-            # Check if we need to change XML of VM.
-            if channel != bool(vmxml.get_agent_channels()):
-                if self.is_alive():
-                    self.destroy()
+            # Check if we need to change XML of VM by checking
+            # whether the agent is existing.
+            is_existing = False
+            ga_channels = vmxml.get_agent_channels()
+            for chnl in ga_channels:
+                name = chnl.find('./target').get('name')
+                try:
+                    path = chnl.find('./source').get('path')
+                except AttributeError:
+                    path = None
+                if (name == target_name and path == source_path):
```

Comment:
Do we need check whether source_path is not None? otherwise if source_path is None, and path is None, it may falsely set is_existing = True.

Reply:
@chunfuwen if source_path==path==None, and path==source_path, then it is the agent xml
I needed. is_existing is True indeed. May I did not get exactly what you mean, please help to check.